pkg/manifests/tls.go

```diff
@@ -36,94 +39,150 @@ func needsNewCert(notBefore, notAfter time.Time, now func() time.Time) bool {
+			fmt.Sprintf("%s@%d", "openshift-cluster-monitoring", time.Now().Unix()),
+			crypto.DefaultCertificateLifetimeInDays,
+		)
 		if err != nil {
-			klog.Warningf("creating a new CA due to error reading CA: %v", err)
-		} else if !needsNewCert(ca.Config.Certs[0].NotBefore, ca.Config.Certs[0].NotAfter, time.Now) {
-			return s, nil
+			return errors.Wrap(err, "error generating self signed CA")
 		}
-	}
 
-	cfg, err := crypto.MakeSelfSignedCAConfig(
-		fmt.Sprintf("%s@%d", "openshift-cluster-monitoring", time.Now().Unix()),
-		crypto.DefaultCertificateLifetimeInDays,
-	)
-	if err != nil {
-		return nil, errors.Wrap(err, "error creating self signed CA")
+		newCA = &crypto.CA{
+			SerialGenerator: &crypto.RandomSerialGenerator{},
+			Config:          newCAConfig,
+		}
+	} else {
+		template := curCA.Config.Certs[0]
+		template.NotAfter = time.Now().Add(time.Duration(crypto.DefaultCertificateLifetimeInDays) * 24 * time.Hour)
+		template.SerialNumber = template.SerialNumber.Add(template.SerialNumber, big.NewInt(1))
+
+		newCACert, err := createCertificate(template, template, template.PublicKey, curCA.Config.Key)
+		if err != nil {
+			return errors.Wrap(err, "error rotating CA")
+		}
+
+		newCA = &crypto.CA{
+			SerialGenerator: &crypto.RandomSerialGenerator{},
+			Config: &crypto.TLSCertificateConfig{
+				Certs: []*x509.Certificate{newCACert},
+				Key:   curCA.Config.Key,
+			},
+		}
 	}
 
-	crt, key, err = cfg.GetPEMBytes()
+	newCABytes, newCAKeyBytes, err := newCA.Config.GetPEMBytes()
 	if err != nil {
-		return nil, err
+		return errors.Wrap(err, "error getting PEM bytes from CA")
 	}
 
-	s.Data["ca.crt"] = crt
-	s.Data["ca.key"] = key
-
-	ca := &crypto.CA{
-		SerialGenerator: &crypto.RandomSerialGenerator{},
-		Config:          cfg,
-	}
+	s.Data["ca.crt"] = newCABytes
+	s.Data["ca.key"] = newCAKeyBytes
 
 	{
-		cfg, err := ca.MakeClientCertificateForDuration(
+		cfg, err := newCA.MakeClientCertificateForDuration(
 			&user.DefaultInfo{
 				Name: "thanos-querier",
 			},
 			time.Duration(crypto.DefaultCertificateLifetimeInDays)*24*time.Hour,
 		)
 		if err != nil {
-			return nil, errors.Wrap(err, "error making thanos querier client certificate")
+			return errors.Wrap(err, "error making thanos querier client certificate")
 		}
+
 		crt, key, err := cfg.GetPEMBytes()
 		if err != nil {
-			return nil, errors.Wrap(err, "error getting PEM bytes for thanos querier client certificate")
+			return errors.Wrap(err, "error getting PEM bytes for thanos querier client certificate")
 		}
 		s.Data["thanos-querier-client.crt"] = crt
 		s.Data["thanos-querier-client.key"] = key
 	}
 
 	{
-		cfg, err := ca.MakeServerCert(
+		cfg, err := newCA.MakeServerCert(
 			sets.NewString("prometheus-grpc"),
 			crypto.DefaultCertificateLifetimeInDays,
 		)
 		if err != nil {
-			return nil, errors.Wrap(err, "error making prometheus-k8s server certificate")
+			return errors.Wrap(err, "error making prometheus-k8s server certificate")
```

Comment:
Nit: What does prometheus-k8s mean to users? 🤔 e.g. the one we deploy in openshift-cluster-monitoring, correct? To users might be helpful to include that?

Reply:
You have a good point, we actually have only two certificate, one server cert (used in thanos ruler and prometheus) and one client cert. I will use these terms instead.